### ISSUE TITLE
conn.close() or something similar?

### ISSUE THREAD

**Issue by DmitryEfimenko** [morris/vinyl-ftp#19]
Sorry for spamming with issues/questions, but I got another one:

So the general flow of my gulp task is that it watches for changes on local files. If file changes it does the following:
* opens connection
* deploys the file
* sets permission on that file
```
var conn = ftp.create({[params removed for brevity]});
gulp.src(['build/' + fileChanged])
        .pipe(conn.dest('../usr/local/www/' + destFolder))
        .pipe(setPermissions(conn, destFolder, fileChanged));
```
... where `setPermissions` essentially does this:
```
var filePermission = getPermission(fileChanged);
conn.chmod('../usr/local/www/' + fileChanged, filePermission, function() {
    console.log('permission', filePermission, 'is set for file', fileChanged);
});
```
What this code misses is closing of the connection. Because of that, eventually gulp stops deploying files with the following error:
```
ERROR : Error: There are too many connections from your internet address.
    at makeError <C:[project path]\node_modules\vinyl-ftp\node_modules\ftp\lib\connection.js:1067:13>
```
Notice, that due to the fact that I'm using conn.chmod, I can't pass options with `{ keep: false }`, to tell vinyl-ftp to close connection after permissions are set. I need a way to close connection somehow differently.
it'd be great if there was an api call `conn.close()`

As usual, any help is appreciated.

**Comment by morris** [comment]
The problem here is that you're not using the vinyl-ftp's internal connection pool. Better put this line: `var conn = ftp.create({[params removed for brevity]});` on top of your gulp file and reuse the connection. This will be much(!) faster and won't stall with too many connections. Another, not recommended approach would be to `conn.disconnect()`. Don't :-)

**Comment by DmitryEfimenko** [comment]
works well. The only thing about this approach is that if parameter `maxConnections` is set to something low (in my case the magic number is 22 or lower), vinyl-ftp errors out saying `Error: FTP disconnected`.
So I set it to a 1000, and it looks like it works fine. I also saw in the README you recommend to set it to something high anyway since there is a good recovery mechanism. If that's the case, why not default it to something very high?

**Comment by morris** [comment]
That's odd, unless you call `disconnect()` manually. Could you paste your complete gulpfile so I can reproduce?

**Comment by DmitryEfimenko** [comment]
sure. General setup is:
* create `app` folder with a bunch of `.js` files in the same folder where this gulpfile is.
* create `build` folder. This is where production ready js files will be copied to. There also will be a watch set up on that folder, which will be responsible for FTPing these files.
* in the `gulpfile` update credentials for `conn`, `maxConnections` to something low, and variable `destFolderRoot` if needed.

---

below is a simplified, but working gulpfile.
```javascript
var gulp = require('gulp');
var gutil = require('gutil');
var ftp = require('vinyl-ftp');
var concat = require('gulp-concat');
var uglify = require('gulp-uglify');
var sourcemaps = require('gulp-sourcemaps');
var eventStream = require('event-stream');

var deployToTest = true;
var destFolderRoot = '../usr/local/www/';

var conn = ftp.create({
  host:     'myIpAddress',
  user:     'myuser',
  password: 'mypasword',
  parallel: 5,
  maxConnections: 1000,
  log: gutil.log
});

var paths = {
  scriptsApp: ['app/**/*.js'],
};

// Rerun the task when a file changes
gulp.task('watch', function() {
  gulp.watch(paths.scriptsApp, ['scripts-app']);

  if(deployToTest) {
    gulp.watch(['build/**'], function(event) {
      var fileChanged = event.path.substring(event.path.indexOf('build') + 6, event.path.length);
      var destFolder = fileChanged.substring(0, fileChanged.lastIndexOf("\\"));
    
      console.log('deploying ' + fileChanged +' to ' + destFolder);
      
      gulp.src(['build/' + fileChanged])
        .pipe(conn.dest(destFolderRoot + destFolder))
        .pipe(setPermissions(conn, destFolder, fileChanged));
    });
  }
});

function getPermForFolder(destFolder) {
  var foldersFor0705 = ['scripts', ''];
  if(foldersFor0705.indexOf(destFolder) > -1) {
    return '0705';
  }
  return '0704';
}

function getFilePermission(destFolder, fileChanged) {
  var foldersFor0707 = ['cgi-bin'];
  if(foldersFor0707.indexOf(destFolder) > -1) {
    return '0707';
  } else {
    return '0704';
  }
}

var folderPermissionsSet = [];
function setPermissions(conn, destFolder, fileChanged) {

  function setFolderPerms(destFolder, cb) {
    // only run permissions on folders once
    if(folderPermissionsSet.indexOf(destFolder) == -1) {
      var folderPermission = getPermForFolder(destFolder);
      conn.chmod(destFolderRoot + destFolder, folderPermission, function(){
        console.log('permission', folderPermission, 'is set for folder', destFolder);
        cb();
      });
      folderPermissionsSet.push(destFolder);
    } else {
      cb();
    }
  }

  function transform(file, cb) {
    setFolderPerms(destFolder, function(){
      var filePermission = getFilePermission(destFolder, fileChanged);
      conn.chmod(destFolderRoot + fileChanged, filePermission, function(){ console.log('permission', filePermission, 'is set for file', fileChanged); });
    })
    cb(null, file);
  }

  return eventStream.map(transform);
}

gulp.task('scripts-app', function() {
  // Minify and copy all JavaScript (except vendor scripts)
  // with sourcemaps all the way down
  gulp.src(paths.scriptsApp)
    .pipe(sourcemaps.init())
    .pipe(uglify())
    .pipe(concat('app.min.js'))
    .pipe(sourcemaps.write())
    .pipe(gulp.dest('build/scripts'));
});

gulp.task('default', ['scripts-app'], function() {
  gulp.start('watch');
});
```

**Comment by DmitryEfimenko** [comment]
Is there any status update on this by any chance?
It is still an issue for me. Sometimes I'll set the `maxConnections` to some high number (ex: 1000), and it'll work for a while, but then I rerun gulp and it stops working with the same error message...

**Comment by morris** [comment]
I finally found some time to reproduce this - it's definitely an issue with the connection pool which I hope to fix soon. Sorry for the delay!

**Comment by DmitryEfimenko** [comment]
Hey, thanks for doing this. However long it takes is fine!

**Comment by morris** [comment]
Fixed in v0.4.4 - it really was an important issue to fix, thank you so much for reporting this :+1: The connection pool is much more stable now, and even faster. Please try it and let me know how this works out.

**Comment by DmitryEfimenko** [comment]
It looks like it works well! Good job!
